### PR TITLE
 Add k8s and qingcloud basic runtime functions

### PR DESCRIPTION
**Pull request by chilianyi**

### CODE REVIEW

File: pkg/cmd/runtime/plugins/k8s/k8s_runtime.go [hunk 1]
```diff
@@ -5,9 +5,25 @@
...
+	"k8s.io/api/extensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apiruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/homedir"
+
+	"fmt"
```
Comment (chai2010) [inline pkg/cmd/runtime/plugins/k8s/k8s_runtime.go]: mv "fmt" to the std package group

File: pkg/cmd/runtime/plugins/k8s/k8s_runtime.go [hunk 2]
```diff
@@ -21,3 +37,167 @@ func (p *K8sRuntime) Name() string { return "k8s" }
 func (p *K8sRuntime) Run(app string, args ...string) error {
 	return errors.New("TODO")
 }
+
+func (p *K8sRuntime) GetDefaultClient() (clientset *kubernetes.Clientset, err error) {
```
Comment (chai2010) [inline pkg/cmd/runtime/plugins/k8s/k8s_runtime.go]: if the `GetDefaultClient` is the private method, it should be `getDefaultClient`, the first letter is lower case.

File: pkg/cmd/runtime/plugins/k8s/k8s_runtime.go [hunk 3]
```diff
@@ -21,3 +37,167 @@ func (p *K8sRuntime) Name() string { return "k8s" }
...
+	}
+
+	decoder := scheme.Codecs.UniversalDeserializer()
+	d := yaml.NewYAMLOrJSONDecoder(strings.NewReader(appConf), 4096)
+	for {
+		ext := apiruntime.RawExtension{}
+		if err := d.Decode(&ext); err != nil {
+			if err == io.EOF {
+				err = nil
+				return clusterId, err
```
Comment (chai2010) [inline pkg/cmd/runtime/plugins/k8s/k8s_runtime.go]: `return clusterId, nil` is better

File: pkg/cmd/runtime/plugins/k8s/k8s_runtime.go (hunk 1 above)
Comment (chilianyi) [reply pkg/cmd/runtime/plugins/k8s/k8s_runtime.go]: fixed

File: pkg/cmd/runtime/plugins/k8s/k8s_runtime.go [hunk 4]
```diff
@@ -21,3 +37,167 @@ func (p *K8sRuntime) Name() string { return "k8s" }
...
+			logger.Errorf("Failed to create cluster with unknow type")
+			return clusterId, fmt.Errorf("Failed to create cluster [%s] with unknow type", clusterId)
+		}
+	}
+
+	return
+}
+
+func (p *K8sRuntime) StartClusters(clusterIds string, shouldWait bool, args ...string) error {
+	if len(args) < 1 {
```
Comment (chai2010) [inline pkg/cmd/runtime/plugins/k8s/k8s_runtime.go]: if `args[0]` must exists, the arguments should be `StartClusters(clusterIds string, shouldWait bool, appConf  string, ...)`

File: pkg/cmd/runtime/plugins/k8s/k8s_runtime.go [hunk 5]
```diff
@@ -21,3 +37,167 @@ func (p *K8sRuntime) Name() string { return "k8s" }
...
+	}
+
+	appConf := args[0]
+	clusterId := strings.Split(clusterIds, ",")[0]
+	_, err := p.CreateCluster(appConf, shouldWait, clusterId)
+	return err
+}
+
+func (p *K8sRuntime) RecoverClusters(clusterIds string, shouldWait bool, args ...string) error {
+	if len(args) < 1 {
```
Comment (chai2010) [inline pkg/cmd/runtime/plugins/k8s/k8s_runtime.go]: ditto

File: pkg/cmd/runtime/plugins/k8s/k8s_runtime.go [hunk 6]
```diff
@@ -21,3 +37,167 @@ func (p *K8sRuntime) Name() string { return "k8s" }
...
+	return err
+}
+
+func (p *K8sRuntime) DeleteClusters(clusterIds string, shouldWait bool, args ...string) error {
+	clientset, err := p.GetDefaultClient()
+	if err != nil {
+		logger.Errorf("Failed to get client when create cluster: %v", err)
+		return err
+	}
+	if len(args) < 1 {
```
Comment (chai2010) [inline pkg/cmd/runtime/plugins/k8s/k8s_runtime.go]: ditto

File: pkg/cmd/runtime/plugins/k8s/k8s_runtime.go [hunk 7]
```diff
@@ -21,3 +37,167 @@ func (p *K8sRuntime) Name() string { return "k8s" }
...
+
+	appConf := args[0]
+	decoder := scheme.Codecs.UniversalDeserializer()
+	d := yaml.NewYAMLOrJSONDecoder(strings.NewReader(appConf), 4096)
+	for {
+		ext := apiruntime.RawExtension{}
+		if err := d.Decode(&ext); err != nil {
+			if err == io.EOF {
+				err = nil
+				return err
```
Comment (chai2010) [inline pkg/cmd/runtime/plugins/k8s/k8s_runtime.go]: return nil

File: pkg/cmd/runtime/plugins/k8s/k8s_runtime.go [hunk 8]
```diff
@@ -21,3 +37,167 @@ func (p *K8sRuntime) Name() string { return "k8s" }
...
+		}
+		obj, _, err := decoder.Decode(ext.Raw, nil, nil)
+		switch o := obj.(type) {
+		case *v1beta1.Deployment:
+			name := p.GetName(clusterIds, o.GetName())
+			deploymentsClient := clientset.ExtensionsV1beta1().Deployments(v1.NamespaceDefault)
+			background := metav1.DeletePropagationBackground
+			err = deploymentsClient.Delete(name, &metav1.DeleteOptions{PropagationPolicy: &background})
+			if err != nil {
+				logger.Errorf("Failed to delete cluster: %v", err)
```
Comment (chai2010) [inline pkg/cmd/runtime/plugins/k8s/k8s_runtime.go]: missing return ?

File: pkg/cmd/runtime/plugins/k8s/k8s_runtime.go [hunk 9]
```diff
@@ -21,3 +37,167 @@ func (p *K8sRuntime) Name() string { return "k8s" }
...
+			logger.Errorf("Failed to delete cluster with unknow type")
+			return fmt.Errorf("Failed to delete cluster [%s] with unknow type", clusterIds)
+		}
+	}
+
+	return err
+}
+
+func (p *K8sRuntime) StopClusters(clusterIds string, shouldWait bool, args ...string) error {
+	if len(args) < 1 {
```
Comment (chai2010) [inline pkg/cmd/runtime/plugins/k8s/k8s_runtime.go]: ditto

File: pkg/cmd/runtime/plugins/k8s/k8s_runtime.go [hunk 10]
```diff
@@ -21,3 +37,167 @@ func (p *K8sRuntime) Name() string { return "k8s" }
...
+		logger.Errorf("Failed to get conf when cease cluster")
+		return fmt.Errorf("Failed to get conf when cease cluster [%s]", clusterIds)
+	}
+
+	appConf := args[0]
+	return p.DeleteClusters(clusterIds, shouldWait, appConf)
+}
+
+func (p *K8sRuntime) GetName(clusterId string, oriName string) string {
+	return (clusterId + "-" + oriName)
```
Comment (chai2010) [inline pkg/cmd/runtime/plugins/k8s/k8s_runtime.go]: `GetName(clusterId, oriName string) string`
`return clusterId + "-" + oriName`

File: pkg/cmd/runtime/plugins/k8s/k8s_runtime_test.go [hunk 11]
```diff
@@ -0,0 +1,82 @@
+package k8s_runtime
```
Comment (chai2010) [inline pkg/cmd/runtime/plugins/k8s/k8s_runtime_test.go]: missing copyright

File: pkg/cmd/runtime/plugins/k8s/k8s_runtime_test.go [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+const appConf = `
```
Comment (chai2010) [inline pkg/cmd/runtime/plugins/k8s/k8s_runtime_test.go]: add t or test or testxxx as prefix is better for global testing variables

File: pkg/utils/uuid.go [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+			initByte = append(initByte, byte(i))
+		case i >= 65 && i < 91:
+			initByte = append(initByte, byte(i))
+		case i >= 97 && i < 123:
+			initByte = append(initByte, byte(i))
+		}
+
+	}
+
+	rand.Seed(time.Now().UnixNano())
```
Comment (chai2010) [inline pkg/utils/uuid.go]: rand.Seed should be in init func, it only called once

File: pkg/utils/uuid.go [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+import (
+	"math/rand"
+	"time"
+)
+
+func GetUuid(l int) string {
+	initByte := []byte{}
+	result := []byte{}
+
+	for i := 48; i < 123; i++ {
```
Comment (chai2010) [inline pkg/utils/uuid.go]: what is 48 and 123 mean?

File: pkg/utils/uuid.go [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+	rand.Seed(time.Now().UnixNano())
+	for i := 0; i < l; i++ {
+		result = append(result, initByte[rand.Intn(len(initByte))])
+	}
+	return string(result)
+}
+
+func GetLowerAndNumUuid(l int) string {
+	initByte := []byte{}
+	result := []byte{}
```
Comment (chai2010) [inline pkg/utils/uuid.go]:
```
var (
  initByte []byte
  result    []byte
)
```

Comment (chilianyi) [reply pkg/utils/uuid.go]: ok

File: pkg/utils/uuid.go (hunk 13 above)
Comment (chilianyi) [reply pkg/utils/uuid.go]: time.Now will change every call

File: pkg/utils/uuid.go (hunk 14 above)
Comment (chilianyi) [reply pkg/utils/uuid.go]: 48 means 0, 122 means z

File: pkg/cmd/runtime/plugins/k8s/k8s_runtime.go (hunk 2 above)
Comment (chilianyi) [reply pkg/cmd/runtime/plugins/k8s/k8s_runtime.go]: ok

File: pkg/cmd/runtime/plugins/k8s/k8s_runtime.go (hunk 3 above)
Comment (chilianyi) [reply pkg/cmd/runtime/plugins/k8s/k8s_runtime.go]: ok

File: pkg/cmd/runtime/plugins/qingcloud/qingcloud_runtime.go [hunk 16]
```diff
@@ -0,0 +1,407 @@
...
+
+func (p *QingcloudRuntime) WaitJobs(jobIds []string, timeout time.Duration, waitInterval time.Duration) error {
+	jobService, err := p.InitJobService()
+	if err != nil {
+		logger.Errorf("Failed to init job service when wait cluster jobs %s: %v", jobIds, err)
+		return err
+	}
+	var wg sync.WaitGroup
+	wg.Add(len(jobIds))
+	done := make(chan error, 1)
```
Comment (chai2010) [inline pkg/cmd/runtime/plugins/qingcloud/qingcloud_runtime.go]: why the done buffer is 1, not len(jobIds) ?

File: pkg/cmd/runtime/plugins/k8s/k8s_runtime.go (hunk 4 above)
Comment (chilianyi) [reply pkg/cmd/runtime/plugins/k8s/k8s_runtime.go]: Here implement runtime, k8s runtime need appConf, but qingcloud runtime not need this parameter

File: pkg/cmd/runtime/plugins/qingcloud/qingcloud_runtime.go [hunk 17]
```diff
@@ -0,0 +1,407 @@
...
+			defer wg.Done()
+			done <- client.WaitJob(jobService, jobId, timeout, waitInterval)
+
+			waitErr, ok := <-done
+			if !ok {
+				return
+			}
+			if waitErr != nil {
+				logger.Errorf("Failed to wait cluster job [%s]: %v", jobId, waitErr)
+				err = waitErr
```
Comment (chai2010) [inline pkg/cmd/runtime/plugins/qingcloud/qingcloud_runtime.go]: it is unsafe for multi goroutine, `err = xxx` is not atomic operate (not sure).

File: pkg/cmd/runtime/plugins/qingcloud/qingcloud_runtime.go [hunk 18]
```diff
@@ -0,0 +1,407 @@
...
+			for _, cluster := range output.ClusterSet {
+				if cluster.TransitionStatus == nil {
+					logger.Errorf("Cluster [%s] transition_status is nil ", clusterIds)
+					return false, nil
+				}
+				if service.StringValue(cluster.TransitionStatus) != "" {
+					return false, nil
+				}
+				if service.StringValue(cluster.Status) == status {
+					continue
```
Comment (chai2010) [inline pkg/cmd/runtime/plugins/qingcloud/qingcloud_runtime.go]: logic is complex

File: pkg/utils/uuid.go (hunk 14 above)
Comment (chai2010) [reply pkg/utils/uuid.go]: `for i := '0'; i < 'z'; i++`

File: pkg/utils/uuid.go (hunk 13 above)
Comment (chai2010) [reply pkg/utils/uuid.go]: rand.Seed only need one random seed, the all next call `rand.XXX` are all random

File: pkg/cmd/runtime/plugins/qingcloud/qingcloud_runtime.go (hunk 17 above)
Comment (chilianyi) [reply pkg/cmd/runtime/plugins/qingcloud/qingcloud_runtime.go]: May generate multiple job, any one of these jobs fail will cause whole failed. 
chan 1 means only one process can write final err to the result err.

File: pkg/cmd/runtime/plugins/k8s/k8s_runtime.go (hunk 7 above)
Comment (chilianyi) [reply pkg/cmd/runtime/plugins/k8s/k8s_runtime.go]: ok

File: pkg/cmd/runtime/plugins/k8s/k8s_runtime.go (hunk 8 above)
Comment (chilianyi) [reply pkg/cmd/runtime/plugins/k8s/k8s_runtime.go]: yes, i will fix it

File: pkg/cmd/runtime/plugins/k8s/k8s_runtime.go (hunk 10 above)
Comment (chilianyi) [reply pkg/cmd/runtime/plugins/k8s/k8s_runtime.go]: ok

File: pkg/cmd/runtime/plugins/k8s/k8s_runtime_test.go (hunk 11 above)
Comment (chilianyi) [reply pkg/cmd/runtime/plugins/k8s/k8s_runtime_test.go]: yes, i will fix it

File: pkg/cmd/runtime/plugins/k8s/k8s_runtime_test.go (hunk 12 above)
Comment (chilianyi) [reply pkg/cmd/runtime/plugins/k8s/k8s_runtime_test.go]: ok

File: pkg/cmd/runtime/plugins/qingcloud/constants.go [hunk 19]
```diff
@@ -0,0 +1,14 @@
+package qingcloud
```
[the remark this answers is not in the document]
Comment (chilianyi) [reply pkg/cmd/runtime/plugins/qingcloud/constants.go]: yes, i will fix it

File: pkg/cmd/runtime/plugins/qingcloud/constants.go [hunk 20]
```diff
@@ -0,0 +1,14 @@
...
+const TIMEOUT_CREATE_CLUSTER = 600 * time.Second
+const TIMEOUT_START_CLUSTER = 600 * time.Second
+const TIMEOUT_STOP_CLUSTER = 600 * time.Second
+const TIMEOUT_DELETE_CLUSTER = 600 * time.Second
+const TIMEOUT_RECOVER_CLUSTER = 600 * time.Second
+const TIMEOUT_CEASE_CLUSTER = 600 * time.Second
+
+const WAIT_INTERVAL = 20 * time.Second
+
+const STATUS_ACTIVE = "active"
```
[the remark this answers is not in the document]
Comment (chilianyi) [reply pkg/cmd/runtime/plugins/qingcloud/constants.go]: ok

File: pkg/cmd/runtime/plugins/qingcloud/qingcloud_runtime.go [hunk 21]
```diff
@@ -0,0 +1,407 @@
...
+)
+
+func init() {
+	runtime.RegisterRuntime(new(QingcloudRuntime))
+}
+
+type QingcloudRuntime struct {
+}
+
+func (p *QingcloudRuntime) InitService() (qingcloudService *service.QingCloudService, err error) {
```
[the remark this answers is not in the document]
Comment (chilianyi) [reply pkg/cmd/runtime/plugins/qingcloud/qingcloud_runtime.go]: private

File: pkg/cmd/runtime/plugins/qingcloud/qingcloud_runtime.go [hunk 22]
```diff
@@ -0,0 +1,407 @@
...
+		return
+	}
+	qingcloudService, err = service.Init(userConf)
+	if err != nil {
+		return
+	}
+	return
+}
+
+func (p *QingcloudRuntime) InitClusterService() (clusterService *service.ClusterService, err error) {
```
[the remark this answers is not in the document]
Comment (chilianyi) [reply pkg/cmd/runtime/plugins/qingcloud/qingcloud_runtime.go]: private

File: pkg/cmd/runtime/plugins/qingcloud/qingcloud_runtime.go (hunk 18 above)
Comment (chilianyi) [reply pkg/cmd/runtime/plugins/qingcloud/qingcloud_runtime.go]: wait the cluster till status transform to the defined status, if transition_status is not "" means the cluster is working,

File: pkg/cmd/runtime/plugins/qingcloud/qingcloud_runtime.go [hunk 23]
```diff
@@ -0,0 +1,407 @@
...
+	return err
+}
+
+func (p *QingcloudRuntime) WaitClustersStatus(clusterService *service.ClusterService, clusterIds string, status string, timeout time.Duration, waitInterval time.Duration) error {
+	return utils.WaitForSpecificOrError(func() (bool, error) {
+		clusters := strings.Split(clusterIds, ",")
+		output, err := clusterService.DescribeClusters(&service.DescribeClustersInput{Clusters: service.StringSlice(clusters)})
+		if err != nil {
+			//network or api error
+			return false, nil
```
Comment (chai2010) [inline pkg/cmd/runtime/plugins/qingcloud/qingcloud_runtime.go]: `return false, err` ?

File: pkg/cmd/runtime/plugins/qingcloud/qingcloud_runtime.go (hunk 18 above)
Comment (chai2010) [reply pkg/cmd/runtime/plugins/qingcloud/qingcloud_runtime.go]:
```go
func (p *QingcloudRuntime) WaitClustersStatus(clusterService *service.ClusterService, clusterIds, status string, timeout time.Duration, waitInterval time.Duration) error {
	return utils.WaitForSpecificOrError(func() (bool, error) {
		clusters := strings.Split(clusterIds, ",")
		output, err := clusterService.DescribeClusters(&service.DescribeClustersInput{Clusters: service.StringSlice(clusters)})
		if err != nil {
			//network or api error
			return false, nil
		}

		if len(output.ClusterSet) != len(clusters) {
			return false, fmt.Errorf("Can not find clusters [%s]", clusterIds)
		} 
		 
		for _, cluster := range output.ClusterSet {
			if cluster.TransitionStatus == nil {
				logger.Errorf("Cluster [%s] transition_status is nil ", clusterIds)
				return false, nil
			}
			if service.StringValue(cluster.TransitionStatus) != "" {
				return false, nil
			}

			if service.StringValue(cluster.Status) != status {
				return false, fmt.Errorf("Cluster [%s] status [%s]", clusterIds, status)
			}

			// continue
		}
		return true, nil
	
	}, timeout, waitInterval)
}
```

File: pkg/utils/uuid.go (hunk 14 above)
Comment (chilianyi) [reply pkg/utils/uuid.go]: ok

File: pkg/cmd/runtime/plugins/qingcloud/qingcloud_runtime.go (hunk 23 above)
Comment (chilianyi) [reply pkg/cmd/runtime/plugins/qingcloud/qingcloud_runtime.go]: network or api error 
retry andwait.

File: pkg/cmd/runtime/plugins/qingcloud/qingcloud_runtime.go (hunk 18 above)
Comment (chilianyi) [reply pkg/cmd/runtime/plugins/qingcloud/qingcloud_runtime.go]: ok

**Pull request merged**